Spurious error appeared suddenly in 1.20.0

**Describe the bug**

Having just upgrade to v1.20.0 and having changed no files in the repo, these errors just started appearing.
These errors were NOT present with v1.19.0.
Note that line 1 of each of the 2 noted files was a standard JavaScript import statement (`import PropTypes from 'prop-types'`)

```
% semgrep -f p/cwe-top-25                     
      
. . . standard info left out here...
                                                                                                                        
  ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╸ 100% 0:00:01                                                                                                                        
[ERROR] Fatal error at line src/components/system-view.jsx:1:
 (Failure "Call AST_utils.with_xxx_equal to avoid this error.")
====[ BEGIN error trace ]====
(Failure "Call AST_utils.with_xxx_equal to avoid this error.")
=====[ END error trace ]=====

[ERROR] Fatal error at line src/components/policy-authoring/aux-container.jsx:1:
 (Failure "Call AST_utils.with_xxx_equal to avoid this error.")
====[ BEGIN error trace ]====
(Failure "Call AST_utils.with_xxx_equal to avoid this error.")
=====[ END error trace ]=====
```

The same issue appears when running it just on either single file in question.
Note the mention of the internal semgrep errors was present with previous semgrep versions; it is just the fatal error that is new with 1.20.0.
```
% semgrep -f p/cwe-top-25 src/components/system-view.jsx
               
               
┌─────────────┐
│ Scan Status │
└─────────────┘
  Scanning 1 file tracked by git with 207 Code rules:
                                                                                                                        
  Language      Rules   Files          Origin      Rules                                                                
 ─────────────────────────────        ───────────────────                                                               
  <multilang>       3       2          Community     207                                                                
  js               41       1                                                                                           
                                                                                                                        
  ━━━━━━━━━━━━━━━━━━━━━━━━━━╸━━━━━━━━━━━━━  67% 0:00:00                                                                                                                        
[ERROR] Fatal error at line src/components/system-view.jsx:1:
 (Failure "Call AST_utils.with_xxx_equal to avoid this error.")
====[ BEGIN error trace ]====
(Failure "Call AST_utils.with_xxx_equal to avoid this error.")
=====[ END error trace ]=====
              
┌──────────────┐
│ Scan Summary │
└──────────────┘
Some files were skipped or only partially analyzed.
  Partially scanned: 1 files only partially analyzed due to parsing or internal Semgrep errors
```

**To Reproduce**
Apparently just need to supply a file that causes Semgrep errors.

**Expected behavior**
No error.

**What is the priority of the bug to you?**

- [ ] P0: blocking your adoption of Semgrep or workflow
- [x] P1: important to fix or quite annoying
- [ ] P2: regular bug that should get fixed

**Environment**
If not using semgrep.dev: are you running off docker, an official binary, a local build?
Both local binary and docker versions do the same thing.

I've just had this error as well, in my CI pipeline, on "line 1" of a plain JavaScript file, which only contains a `use strict` statement.

```log
[ERROR] Fatal error at line /src/web/modules/custom/jjj_utils/js/jjj_utils.js:1:
 (Failure "Call AST_utils.with_xxx_equal to avoid this error.")
====[ BEGIN error trace ]====
(Failure "Call AST_utils.with_xxx_equal to avoid this error.")
=====[ END error trace ]=====
```

The start of the file:

```javascript
'use strict';

/**
 * @file
 * Contains global functions we can call from other modules.
 */

(function($) {
  ...
```

--- a/changelog.d/gh-7694.fixed
+++ b/changelog.d/gh-7694.fixed
@@ -1,0 +1,2 @@
+taint-mode: Fixed bug in taint labels that was causing some fatal errors:
+> Failure "Call AST_utils.with_xxx_equal to avoid this error."

--- a/semgrep.yml
+++ b/semgrep.yml
@@ -365,3 +365,19 @@ rules:
     fix: Common.tl_exn "unexpected empty list"
     languages: [ocaml]
     severity: WARNING
+
+  - id: no-metavariable-equal-xyz
+    patterns:
+      - pattern: Metavariable.$F
+      - pattern-not: Metavariable.equal_mvar
+      - metavariable-regex:
+          metavariable: $F
+          regex: equal_.*
+    message: >-
+      Don't use 'Metavariable.$F' directly, see 'AST_utils'. Instead use
+      either 'Metavariable.Structural.$F' or 'Metavariable.Referential.$F'.
+      You can also consider replacing `equal_mvalue` with
+      'Matching_generic.equal_ast_bound_code'.
+    fix: Metavariable.Structural.$F
+    languages: [ocaml]
+    severity: WARNING

--- a/src/tainting/Dataflow_tainting.ml
+++ b/src/tainting/Dataflow_tainting.ml
@@ -367,7 +367,7 @@ let is_func_sink_with_focus taint_sink =
       true
   | __else__ -> false
 
-let unify_mvars_sets mvars1 mvars2 =
+let unify_mvars_sets env mvars1 mvars2 =
   let xs =
     List.fold_left
       (fun xs (mvar, mval) ->
@@ -375,7 +375,7 @@ let unify_mvars_sets mvars1 mvars2 =
         match List.assoc_opt mvar mvars2 with
         | None -> Some ((mvar, mval) :: xs)
         | Some mval' ->
-            if Metavariable.equal_mvalue mval mval' then
+            if Matching_generic.equal_ast_bound_code env.options mval mval' then
               Some ((mvar, mval) :: xs)
             else None)
       (Some []) mvars1
@@ -393,7 +393,10 @@ let sink_biased_union_mvars source_mvars sink_mvars =
   in
   Some (source_mvars' @ sink_mvars)
 
-let merge_source_mvars bindings =
+(* Takes the bindings of multiple taint sources and filters the bindings ($MVAR, MVAL)
+ * such that either $MVAR is bound by a single source, or all MVALs bounds to $MVAR
+ * can be unified. *)
+let merge_source_mvars env bindings =
   let flat_bindings = List.concat bindings in
   let bindings_tbl =
     flat_bindings
@@ -415,8 +418,10 @@ let merge_source_mvars bindings =
              *)
              Hashtbl.replace bindings_tbl mvar (Some mval)
          | Some (Some mval') ->
-             if not (Metavariable.equal_mvalue mval mval') then
-               Hashtbl.remove bindings_tbl mvar);
+             if
+               not
+                 (Matching_generic.equal_ast_bound_code env.options mval mval')
+             then Hashtbl.remove bindings_tbl mvar);
   (* After this, the only surviving bindings should be those where
      there was no conflict between bindings in different sources.
   *)
@@ -438,7 +443,7 @@ let merge_source_sink_mvars env source_mvars sink_mvars =
      * `pattern-sinks` as independent. We keep this option mainly for
      * backwards compatibility, it may be removed later on if no real use
      * is found. *)
-    unify_mvars_sets source_mvars sink_mvars
+    unify_mvars_sets env source_mvars sink_mvars
   else
     (* The union of both sets, but taking the sink mvars in case of collision. *)
     sink_biased_union_mvars source_mvars sink_mvars
@@ -560,7 +565,7 @@ let findings_of_tainted_sink env taints_with_traces (sink : T.sink) :
                     }))
       else
         match
-          taints_and_bindings |> Common.map snd |> merge_source_mvars
+          taints_and_bindings |> Common.map snd |> merge_source_mvars env
           |> merge_source_sink_mvars env sink_pm.PM.env
         with
         | None -> []

--- a/src/tainting/dune
+++ b/src/tainting/dune
@@ -5,6 +5,7 @@
  (libraries
    pfff-lang_GENERIC-analyze
    semgrep_core
+   semgrep.matching
  )
  (preprocess (pps ppx_deriving.show ppx_profiling))
 )

--- a/tests/rules/no_fatal_error_with_xxx_equal.js
+++ b/tests/rules/no_fatal_error_with_xxx_equal.js
@@ -1,0 +1,8 @@
+// https://github.com/returntocorp/semgrep/issues/7694
+// We don't want to get: "Fatal error": (Failure "Call AST_utils.with_xxx_equal to avoid this error.")
+$('.invalid', c).foobar(
+  function() {
+      //ruleid: test
+      $l = $t.baz('<foo for="' + id + '"></foo>');
+  }
+);

--- a/tests/rules/no_fatal_error_with_xxx_equal.yaml
+++ b/tests/rules/no_fatal_error_with_xxx_equal.yaml
@@ -1,0 +1,22 @@
+rules:
+  - id: test
+    message: Test
+    languages:
+      - javascript
+      - typescript
+    severity: WARNING
+    mode: taint
+    pattern-sources:
+      - label: TEST
+        by-side-effect: true
+        patterns:
+          - pattern-either:
+              - pattern: $A($SOURCE)
+              - pattern: $A.$SANITIZE($SOURCE)
+          - focus-metavariable: $SOURCE
+          - metavariable-regex:
+              metavariable: $A
+              regex: (?i)(.*valid)
+    pattern-sinks:
+      - requires: TEST
+        pattern: '"$HTMLSTR" + $EXPR'

